tests/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing.UnitTests/CodeFixIterationTests.cs

```diff
@@ -0,0 +1,355 @@
+        }
+
+        /// <summary>
+        /// Reports a diagnostic on any integer literal token with a value less than five.
+        /// </summary>
+        [DiagnosticAnalyzer(LanguageNames.CSharp)]
+        private class LiteralUnderFiveAnalyzer : DiagnosticAnalyzer
+        {
+            internal static readonly DiagnosticDescriptor Descriptor =
+                new DiagnosticDescriptor("LiteralUnderFive", "title", "message", "category", DiagnosticSeverity.Warning, isEnabledByDefault: true);
+
+            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+
+            public override void Initialize(AnalysisContext context)
+            {
+                context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+                context.RegisterSyntaxNodeAction(HandleNumericLiteralExpression, SyntaxKind.NumericLiteralExpression);
+            }
+
+            private void HandleNumericLiteralExpression(SyntaxNodeAnalysisContext context)
+            {
+                var node = (LiteralExpressionSyntax)context.Node;
+                if (int.TryParse(node.Token.ValueText, out var value) && value < 5)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, node.Token.GetLocation()));
+                }
+            }
+        }
+
+        [ExportCodeFixProvider(LanguageNames.CSharp)]
+        [PartNotDiscoverable]
+        private class IncrementFix : CodeFixProvider
+        {
+            public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(LiteralUnderFiveAnalyzer.Descriptor.Id);
+
+            public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+            public override Task RegisterCodeFixesAsync(CodeFixContext context)
+            {
+                foreach (var diagnostic in context.Diagnostics)
+                {
+                    context.RegisterCodeFix(
+                        CodeAction.Create(
+                            "LiteralUnderFive",
+                            cancellationToken => CreateChangedDocument(context.Document, diagnostic.Location.SourceSpan, cancellationToken),
+                            nameof(IncrementFix)),
+                        diagnostic);
+                }
+
+                return Task.CompletedTask;
+            }
+
+            private async Task<Document> CreateChangedDocument(Document document, TextSpan sourceSpan, CancellationToken cancellationToken)
+            {
+                var tree = await document.GetSyntaxTreeAsync(cancellationToken);
+                var root = await tree.GetRootAsync(cancellationToken);
+                var token = root.FindToken(sourceSpan.Start);
+                var replacement = int.Parse(token.ValueText) + 1;
+                var newToken = SyntaxFactory.Literal(token.LeadingTrivia, " " + replacement.ToString(), replacement, token.TrailingTrivia);
```

Comment:
📝 The addition of a space here is due to the way we detect if the application of a code fix changed the diagnostics. The current algorithm assumes that the same diagnostic (by ID) appearing in the same location means the code fix iteration process has reached a steady-state. Obviously this is not correct for the sample diagnostic/fix created for this test, and it's reasonable to believe other analyzers could behave in a similar manner. I'll file a follow-up issue to address this.